jerry-core/api/jerry-snapshot.c

```diff
@@ -651,6 +732,259 @@ jerry_exec_snapshot (const uint32_t *snapshot_p, /**< snapshot */
+      {
+        if (literal_start_p[i] != JMEM_CP_NULL)
+        {
+          literal_start_p[i] = ecma_snapshot_get_literal (literal_base_p, number_base_p, literal_start_p[i]);
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* scan_snapshot_functions */
+
+/**
+ * Update all literal offsets in a snapshot data.
+ */
+static void
+update_literal_offsets (uint8_t *buffer_p, /**< snapshot buffer start */
+                        const uint8_t *buffer_end_p, /**< snapshot buffer start */
+                        lit_mem_to_snapshot_id_map_entry_t *lit_map_p) /**< literal map */
+{
+  JERRY_ASSERT (buffer_end_p > buffer_p);
+
+  do
+  {
+    ecma_compiled_code_t *bytecode_p = (ecma_compiled_code_t *) buffer_p;
+    uint32_t code_size = ((uint32_t) bytecode_p->size) << JMEM_ALIGNMENT_LOG;
+
+    if (bytecode_p->status_flags & CBC_CODE_FLAGS_FUNCTION)
+    {
+      jmem_cpointer_t *literal_start_p;
+      uint32_t const_literal_end;
+
+      if (bytecode_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint16_arguments_t));
+
+        cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+      else
+      {
+        literal_start_p = (jmem_cpointer_t *) (buffer_p + sizeof (cbc_uint8_arguments_t));
+
+        cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) buffer_p;
+        const_literal_end = args_p->const_literal_end;
+      }
+
+      for (uint32_t i = 0; i < const_literal_end; i++)
+      {
+        if (literal_start_p[i] != JMEM_CP_NULL)
+        {
+          lit_mem_to_snapshot_id_map_entry_t *current_p = lit_map_p;
+
+          while (current_p->literal_id != literal_start_p[i])
+          {
+            current_p++;
+          }
+
+          literal_start_p[i] = current_p->literal_offset;
+        }
+      }
+    }
+
+    buffer_p += code_size;
+  }
+  while (buffer_p < buffer_end_p);
+} /* update_literal_offsets */
+
+#endif /* JERRY_ENABLE_SNAPSHOT_SAVE */
+
+/**
+ * Merge multiple snapshots into a single buffer
+ *
+ * @return length of merged snapshot file
+ *         0 on error
+ */
+size_t
+jerry_merge_snapshots (const uint32_t **inp_buffers_p, /**< array of (pointers to start of) input buffers */
+                       size_t *inp_buffer_sizes_p, /**< array of input buffer sizes */
+                       size_t number_of_snapshots, /**< number of snapshots */
+                       uint32_t *out_buffer_p, /**< output buffer */
+                       size_t out_buffer_size, /**< output buffer size */
+                       const char **error_p) /**< error description */
+{
+#ifdef JERRY_ENABLE_SNAPSHOT_SAVE
+  uint32_t number_of_funcs = 0;
+  uint32_t merged_configuration = 0;
+  size_t functions_size = sizeof (jerry_snapshot_header_t);
+
+  if (number_of_snapshots < 2)
```

Comment:
Actually, would it do any harm to allow a single snapshot to be "merged" (even if a strange corner case)? If I understand the below code correctly, it would just create a copy of the input snapshot. So perhaps the only case to be excluded is `number_of_snapshots == 0`?

Reply:
(Footnote: actually, I could imagine `number_of_snapshots == 0`, too, producing and empty snapshot, or perhaps a snapshot with a header only, just to make the function totally general. But I see some `JERRY_ASSERT (header_p->number_of_funcs > 0);`, `JERRY_ASSERT (number_of_funcs > 0);`, and `(number_of_funcs - 1)` constructs below, and I still did not get my head around them whether they are artificial constraints or they are really needed.)

Reply:
This can be a future request.

Reply:
(Note: merging merged snapshots could be done at some point).